### PR TITLE
libedgetpu: init at 0-unstable-2024-03-14

### LINKED ISSUE #188719: libedgetpu: package and provide nixos module
**Project description**

It would be nice to be able to use the Google Coral Edge TPU Usb device with NixOS.

I think it's mostly a matter of building libedgetpu (and maybe tensorflow?) and then bundling the udev rules with it?

**Metadata**

* homepage URL:  https://github.com/google-coral/libedgetpu
* source URL: https://github.com/google-coral/libedgetpu
* license: apache 2
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by frenetic00**
## Description of changes

Adds package for [libedgetpu](https://github.com/google-coral/libedgetpu) which is the user space driver for Coral devices.
Fixes #188719 

The package is based on derivations that have been posted in issue #188719
The derivation applies a [patch](https://github.com/google-coral/libedgetpu/pull/66) to the Makefile which allows the library to be built with newer versions of Tensorflow and Flatbuffers.

This as been tested and works with [Frigate 0.13.2](https://github.com/NixOS/nixpkgs/blob/nixos-24.05/nixos/modules/services/video/frigate.nix)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    };
+  });
+in
+stdenv.mkDerivation rec {
+  pname = "libedgetpu";
+  version = "0-unstable-2024-03-14";
+
+  src = fetchFromGitHub {
+    owner = "google-coral";
+    repo = pname;
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:32]:
```suggestion
    repo = "libedgetpu";
```

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  ];
+
+  NIX_CXXSTDLIB_COMPILE = "-std=c++17";
+
+  TFROOT = "${fetchFromGitHub {
+    owner = "tensorflow";
+    repo = "tensorflow";
+    rev = "v2.16.1";
+    sha256 = "sha256-UPvK5Kc/FNVJq3FchN5IIBBObvcHtAPVv0ARzWzA35M=";
+  }}";
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:64]:
```suggestion
  TFROOT = fetchFromGitHub {
    owner = "tensorflow";
    repo = "tensorflow";
    rev = "v2.16.1";
    sha256 = "sha256-UPvK5Kc/FNVJq3FchN5IIBBObvcHtAPVv0ARzWzA35M=";
  };
```

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+  enableParallelBuilding = true;
+
+  installPhase = ''
+    mkdir -p $out/lib
+    cp out/direct/k8/libedgetpu.so.1.0 $out/lib
+    ln -s $out/lib/libedgetpu.so.1.0 $out/lib/libedgetpu.so.1
+    mkdir -p $out/lib/udev/rules.d
+    cp debian/edgetpu-accelerator.rules $out/lib/udev/rules.d/99-edgetpu-accelerator.rules
+  '';
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:74]:
```suggestion
  installPhase = ''
    runHook preInstall
    install -Dm555 out/direct/k8/libedgetpu.so.1.0 -t $out/lib
    ln -s $out/lib/libedgetpu.so.1.0 $out/lib/libedgetpu.so.1
    install -Dm444 debian/edgetpu-accelerator.rules $out/lib/udev/rules.d/99-edgetpu-accelerator.rules
    runHook postInstall
  '';
```

Comment (Aleksanaa) [reply pkgs/by-name/li/libedgetpu/package.nix:74]: Are there generated headers and pkg-config files?

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  meta = with lib; {
+    homepage = "https://github.com/google-coral/libedgetpu";
+    description = "Userspace level runtime driver for Coral devices";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ frenetic00 ];
+  };
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:81]:
```suggestion
  meta = {
    homepage = "https://github.com/google-coral/libedgetpu";
    description = "Userspace level runtime driver for Coral devices";
    license = lib.licenses.asl20;
    maintainers = with lib.maintainers; [ frenetic00 ];
    platforms = lib.platforms.linux;
  };
```

**Review by Aleksanaa — APPROVED** [review]

File: pkgs/by-name/li/libedgetpu/package.nix (hunk 3 above)
Comment (frenetic00) [reply pkgs/by-name/li/libedgetpu/package.nix]: @Aleksanaa Thanks for your suggestions.

There aren't any pkg-config files from what I can see.

Regarding the headers, I've checked how the library is packaged for Debian and I've added a line in `installPhase` to install the relevant header files for the library.

**Review by Aleksanaa — APPROVED** [review]

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    configureFlags = (oldAttrs.configureFlags or [ ]) ++ [ "--enable-shared" ];
+    src = fetchFromGitHub {
+      owner = "google";
+      repo = "flatbuffers";
+      rev = "v${version}";
+      sha256 = "sha256-e+dNPNbCHYDXUS/W+hMqf/37fhVgEGzId6rhP3cToTE=";
+    };
+  });
+in
+stdenv.mkDerivation rec {
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:26]:
```suggestion
stdenv.mkDerivation {
```

Unused rec

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  flatbuffers_23_5_26 = flatbuffers.overrideAttrs (oldAttrs: rec {
+    version = "23.5.26";
+    cmakeFlags = (oldAttrs.cmakeFlags or [ ]) ++ [ "-DFLATBUFFERS_BUILD_SHAREDLIB=ON" ];
+    NIX_CXXSTDLIB_COMPILE = "-std=c++17";
+    configureFlags = (oldAttrs.configureFlags or [ ]) ++ [ "--enable-shared" ];
+    src = fetchFromGitHub {
+      owner = "google";
+      repo = "flatbuffers";
+      rev = "v${version}";
+      sha256 = "sha256-e+dNPNbCHYDXUS/W+hMqf/37fhVgEGzId6rhP3cToTE=";
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:22]:
```suggestion
      hash = "sha256-e+dNPNbCHYDXUS/W+hMqf/37fhVgEGzId6rhP3cToTE=";
```

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+in
+stdenv.mkDerivation rec {
+  pname = "libedgetpu";
+  version = "0-unstable-2024-03-14";
+
+  src = fetchFromGitHub {
+    owner = "google-coral";
+    repo = "libedgetpu";
+    rev = "e35aed18fea2e2d25d98352e5a5bd357c170bd4d";
+    sha256 = "sha256-SabiFG/EgspiCFpg8XQs6RjFhrPPUfhILPmYQQA1E2w=";
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:34]:
```suggestion
    hash = "sha256-SabiFG/EgspiCFpg8XQs6RjFhrPPUfhILPmYQQA1E2w=";
```

File: pkgs/by-name/li/libedgetpu/package.nix [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    flatbuffers_23_5_26
+  ];
+
+  NIX_CXXSTDLIB_COMPILE = "-std=c++17";
+
+  TFROOT = fetchFromGitHub {
+    owner = "tensorflow";
+    repo = "tensorflow";
+    rev = "v2.16.1";
+    sha256 = "sha256-UPvK5Kc/FNVJq3FchN5IIBBObvcHtAPVv0ARzWzA35M=";
```
Comment (Aleksanaa) [inline pkgs/by-name/li/libedgetpu/package.nix:63]:
```suggestion
    hash = "sha256-UPvK5Kc/FNVJq3FchN5IIBBObvcHtAPVv0ARzWzA35M=";
```

**Review by Aleksanaa — APPROVED** [review]
Please squash into two commits:

1. maintainers: add frenetic00
2. libedgetpu: init at 0-unstable-2024-03-14

**Comment by VTimofeenko** [comment]
Works perfectly with frigate from 24.05 and usb-c coral

**Comment by naggie** [comment]
Thank you so much @frenetic00 @Aleksanaa !

**Pull request merged**